File: src/leetcode/m150/EvaluateReversePolishNotation.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m150;

import java.util.Stack;

/**
 * LeetCode 150
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/09/30 12:51
 */
public class EvaluateReversePolishNotation {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            switch (token) {
                case "+":
                    stack.push(stack.pop() + stack.pop());
                    break;
                case "-": {
                    int latter = stack.pop();
                    int former = stack.pop();
                    stack.push(former - latter);
                    break;
                }
                case "*":
                    stack.push(stack.pop() * stack.pop());
                    break;
                case "/": {
                    int latter = stack.pop();
                    int former = stack.pop();
                    stack.push(former / latter);
                    break;
                }
                default:
                    stack.push(Integer.parseInt(token));
                    break;
            }
        }
        return stack.pop();
    }
}

File: src/leetcode/m150/EvaluateReversePolishNotation.go
package m150

import "strconv"

// EvaluateReversePolishNotation LeetCode 150
type EvaluateReversePolishNotation interface {
	evalRPN(tokens []string) int
}

type Solution1 struct{}

func (s Solution1) evalRPN(tokens []string) int {
	stack := make([]string, 0)
	operators := map[string]struct{}{
		"+": {},
		"-": {},
		"*": {},
		"/": {},
	}
	for _, token := range tokens {
		if _, exists := operators[token]; exists {
			first := stack[len(stack)-2]
			second := stack[len(stack)-1]
			stack = stack[:len(stack)-2]
			firstNum, _ := strconv.Atoi(first)
			secondNum, _ := strconv.Atoi(second)
			if token == "+" {
				stack = append(stack, strconv.Itoa(firstNum+secondNum))
			} else if token == "-" {
				stack = append(stack, strconv.Itoa(firstNum-secondNum))
			} else if token == "*" {
				stack = append(stack, strconv.Itoa(firstNum*secondNum))
			} else {
				stack = append(stack, strconv.Itoa(firstNum/secondNum))
			}
		} else {
			stack = append(stack, token)
		}
	}
	res, _ := strconv.Atoi(stack[0])
	return res
}

type Solution2 struct{}

func (s Solution2) evalRPN(tokens []string) int {
	stack := make([]int, 0)
	for _, token := range tokens {
		if token == "+" {
			first, second := popTwoElements(&stack)
			stack = append(stack, first+second)
		} else if token == "-" {
			first, second := popTwoElements(&stack)
			stack = append(stack, first-second)
		} else if token == "*" {
			first, second := popTwoElements(&stack)
			stack = append(stack, first*second)
		} else if token == "/" {
			first, second := popTwoElements(&stack)
			stack = append(stack, first/second)
		} else {
			num, _ := strconv.Atoi(token)
			stack = append(stack, num)
		}
	}
	return stack[0]
}

func popTwoElements(stack *[]int) (int, int) {
	s := *stack
	first := s[len(s)-2]
	second := s[len(s)-1]
	*stack = s[:len(s)-2]
	return first, second
}
